### PR TITLE
Topic/urdf memory leak

### LINKED ISSUE #2077: Memory leak when using buildGeomFromUrdfString
Line
https://github.com/stack-of-tasks/pinocchio/blob/4512214cf08e765a28542c75bf6378e934622a38/bindings/python/parsers/urdf/geometry.cpp#L174
ties the lifetime of the return value to the lifetime of the optional argument "geometry_model". As a consequence, when the argument is not specified by the caller (python side), the value returned by the function is not tied to any python object and therefore the return value, initially dynamically allocated by buildGeomFromUrdfString, is never freed.

@PierreMinotti FYI

### PR DESCRIPTION
**Pull request by jorisv**
close #2077

### CODE REVIEW

File: unittest/python/CMakeLists.txt [hunk 1]
```diff
@@ -2,6 +2,45 @@
 # Copyright (c) 2015-2021 CNRS INRIA
```
Comment (jcarpent) [inline unittest/python/CMakeLists.txt:2]:
```suggestion
# Copyright (c) 2015-2023 CNRS INRIA
```

**Review by jcarpent — DISMISSED** [review]

**Review by jcarpent — DISMISSED** [review]

File: bindings/python/parsers/urdf/geometry.cpp [hunk 2]
```diff
@@ -139,10 +139,20 @@ namespace pinocchio
...
           return bp::return_internal_reference<owner_arg>::postcall(args_, result);
-        return result;
+        // If owner_arg doesn't exist, then Python will have to manage the result lifetime
+        bp::extract<GeometryModel*> geom_model_extract(result);
+        if (geom_model_extract.check())
+        {
+          return bp::to_python_indirect<GeometryModel, bp::detail::make_owning_holder>()(geom_model_extract());
+        }
+        // If returned value is not a GeometryModel*, then raise an error
+        PyErr_SetString(PyExc_RuntimeError, "pinocchio::python::return_value_policy only work on GeometryModel* data type");
```
Comment (ManifoldFR) [inline bindings/python/parsers/urdf/geometry.cpp:154]: "only work**s**"

Comment (ManifoldFR) [reply bindings/python/parsers/urdf/geometry.cpp:154]: the rest looks right to me, `make_owning_holder` seems to create a `unique_ptr` so it shouldn't leak

**Review by ManifoldFR — APPROVED** [review]

File: unittest/python/utils.py [hunk 3]
```diff
@@ -15,9 +15,9 @@ def test_se3ToXYZQUAT_XYZQUATToSe3(self):
         self.assertApprox(pin.SE3ToXYZQUAT(m).T, [1., 2., 3., sqrt(2) / 2, 0, 0, sqrt(2) / 2])
         self.assertApprox(pin.XYZQUATToSE3([1., 2., 3., sqrt(2) / 2, 0, 0, sqrt(2) / 2]), m)
         self.assertApprox(pin.XYZQUATToSE3(np.array([1., 2., 3., sqrt(2) / 2, 0, 0, sqrt(2) / 2])), m)
-        with self.assertRaisesRegex(ValueError, "Wrong size: .*"):
+        with self.assertRaises(ValueError):
             pin.XYZQUATToSE3([1., 2., 3.])
-        with self.assertRaisesRegex(ValueError, "Wrong size: .*"):
```
[the remark this answers is not in the document]
Comment (ManifoldFR) [reply unittest/python/utils.py:20]: Very sad

[the remark this answers is not in the document]
Comment (olivier-stasse) [reply unittest/python/utils.py:20]: I am  sure you already know but Melodic is end of life.
Would that make sense to have like in ROS branches for releases where we do not port new stuff ?

[the remark this answers is not in the document]
Comment (olivier-stasse) [reply unittest/python/utils.py:20]: But I agree that fixing this memory leak is nice also for older releases.

**Review by ManifoldFR — APPROVED** [review]

**Pull request merged**